Sources/TuistKit/Cache/CacheController.swift

```diff
@@ -237,24 +237,44 @@ final class CacheController: CacheControlling {
 
         let graphTraverser = GraphTraverser(graph: graph)
 
-        return try topologicalSort(
+        let graph = try topologicalSort(
             Array(graphTraverser.allTargets()),
             successors: {
                 Array(graphTraverser.directTargetDependencies(path: $0.path, name: $0.target.name))
             }
         )
-        .filter { target in
-            guard let hash = hashesByCacheableTarget[target] else { return false }
-            let cacheExists = try cache.exists(hash: hash).toBlocking().first() ?? false
-            if cacheExists {
-                logger.pretty("The target \(.bold(.raw(target.target.name))) with hash \(.bold(.raw(hash))) and type \(cacheOutputType.description) is already in the cache. Skipping...")
+
+        var targetCacheExists = [String: Bool]()
+        let group = DispatchGroup()
+
+        for target in graph {
+            group.enter()
+
+            DispatchQueue.global().async {
```

Comment:
I would prefer if we try a different implementation, ideally, with RxSwift, we may be able to treat it as a stream of checks and should be fairly easy to implement, will try to see if I can suggest something more concrete later today (or tomorrow).
Otherwise, there is also a `map(context: .parallel)` method that wraps what you just wrote here.

Reply:
GTM. As I said, I tried but I'm not really used to RxSwift.
I managed to create an array of `Single<Bool>` but then I struggled in understanding how to put everything together.

If you can send me some pseudocode or some resources about which methods I should check I can give it another shot

Reply:
In the meanwhile, I pushed the code that uses `map(.concurrent)`. The previous code wasn't thread-safe and I experienced some crashes